libminifi/include/core/ProcessSession.h

```diff
@@ -155,16 +156,25 @@ class ProcessSession : public ReferenceContainer {
     std::shared_ptr<FlowFile> snapshot;
   };
 
+  using Relationships = utils::ValueIdProvider<Relationship>;
+
+  Relationships relationships_;
```

Comment:
Why isn't this a simple `unordered_set`, while references to `Relationship`s with `id_type`, could just be raw pointers to items stored here? Invalid ones are `nullptr`, if necessary. This could eliminate the need for `ValueIdProvider`.

Reply:
could be, but how we achieve a lower memory usage is not really the concern of `ProcessSession`, this `ValueIdProvider` (rename pending) hides this away, `id_type` could be a raw pointer, `INVALID_ID` could be redefined, and we wouldn't have to touch `ProcessSession` for such a change

Reply:
If we used raw pointers, the only member in `ValueCompressor` would be an `unordered_set`. `compress` would become just an `insert`, `decompress` would just become a pointer dereferencing. Is it still worth creating a class for that, basically a wrapper around a standard container?

Reply:
I would rather we use ids, we could configure the id (compressed_type) to be smaller than 8 bytes, we also can't accidentally dereference an invalid pointer (although we could decompress what was compressed with another compressor, that would not be UB at least)

Reply:
We only instantiate `ValueCompressor`  with id being 8 bytes, so this doesn't apply now. Also this member should be private as no subclasses use it, so invalid pointer dereferencing is not possible (by someone else) if usage is consistent within this class.

Reply:
removed `ValueCompressor`, using a pointer into a cache set